File: java/io/github/anaruto/libbbf/BBFReader.java
package io.github.anaruto.libbbf;

import java.nio.ByteBuffer;

public class BBFReader implements AutoCloseable {
    static {
        System.loadLibrary("bbf");
    }

    private final long handle;

    public BBFReader(String filePath) {
        this.handle = openNative(filePath);
        if (this.handle == 0) {
            throw new RuntimeException("Failed to open BBF container at " + filePath);
        }
    }

    public BBFReader(byte[] bytes) {
        this.handle = openBytesNative(bytes);
        if (this.handle == 0) {
            throw new RuntimeException("Failed to open BBF container from bytes");
        }
    }

    @Override
    public void close() {
        closeNative(handle);
    }

    public int getVersion() {
        return getVersion(handle);
    }

    public int getHeaderFlags() {
        return getHeaderFlags(handle);
    }

    public int getAlignment() {
        return getAlignment(handle);
    }

    public int getReamSize() {
        return getReamSize(handle);
    }

    public long getAssetCount() {
        return getAssetCount(handle);
    }

    public long getPageCount() {
        return getPageCount(handle);
    }

    public long getSectionCount() {
        return getSectionCount(handle);
    }

    public long getMetaCount() {
        return getMetaCount(handle);
    }

    public BBFMeta getMeta(int index) {
        String key = getMetaKey(handle, index);
        String val = getMetaValue(handle, index);
        String parent = getMetaParent(handle, index);
        return new BBFMeta(key, val, parent);
    }

    public BBFSection getSection(int index) {
        String title = getSectionTitle(handle, index);
        long start = getSectionStartIndex(handle, index);
        String parent = getSectionParent(handle, index);
        return new BBFSection(title, start, parent);
    }

    public long getPageAssetIndex(int index) {
        return getPageAssetIndex(handle, index);
    }

    public int getPageFlags(int index) {
        return getPageFlags(handle, index);
    }

    public BBFAsset getAsset(int index) {
        long offset = getAssetFileOffset(handle, index);
        long size = getAssetFileSize(handle, index);
        int flags = getAssetFlags(handle, index);
        int type = getAssetType(handle, index);
        byte[] hash = getAssetHash(handle, index);
        return new BBFAsset(offset, hash, size, flags, type);
    }

    public byte[] getAssetData(int index) {
        return getAssetData(handle, index);
    }

    public int readAssetDataDirect(int index, ByteBuffer buffer, int offset, int len) {
        if (!buffer.isDirect()) {
            throw new IllegalArgumentException("ByteBuffer must be a direct buffer");
        }
        return readAssetDataDirect(handle, index, buffer, offset, len);
    }

    public boolean verifyFooterHash() {
        return verifyFooterHash(handle);
    }

    public boolean verifyAssetHash(int index) {
        return verifyAssetHash(handle, index);
    }

    // JNI Declarations
    private static native long openNative(String filePath);
    private static native long openBytesNative(byte[] bytes);
    private static native void closeNative(long handle);

    private static native int getVersion(long handle);
    private static native int getHeaderFlags(long handle);
    private static native int getAlignment(long handle);
    private static native int getReamSize(long handle);

    private static native long getAssetCount(long handle);
    private static native long getPageCount(long handle);
    private static native long getSectionCount(long handle);
    private static native long getMetaCount(long handle);

    private static native String getMetaKey(long handle, int index);
    private static native String getMetaValue(long handle, int index);
    private static native String getMetaParent(long handle, int index);

    private static native String getSectionTitle(long handle, int index);
    private static native long getSectionStartIndex(long handle, int index);
    private static native String getSectionParent(long handle, int index);

    private static native long getPageAssetIndex(long handle, int index);
    private static native int getPageFlags(long handle, int index);

    private static native long getAssetFileOffset(long handle, int index);
    private static native long getAssetFileSize(long handle, int index);
    private static native int getAssetFlags(long handle, int index);
    private static native int getAssetType(long handle, int index);
    private static native byte[] getAssetHash(long handle, int index);

    private static native byte[] getAssetData(long handle, int index);
    private static native int readAssetDataDirect(long handle, int index, ByteBuffer buffer, int offset, int len);

    private static native boolean verifyFooterHash(long handle);
    private static native boolean verifyAssetHash(long handle, int index);
}

File: kotlin/io/github/anaruto/libbbf/BBFReader.kt
package io.github.anaruto.libbbf

import java.nio.ByteBuffer

class BBFReader : AutoCloseable {
    private val handle: Long

    constructor(filePath: String) {
        handle = openNative(filePath)
        if (handle == 0L) {
            throw RuntimeException("Failed to open BBF container at $filePath")
        }
    }

    constructor(bytes: ByteArray) {
        handle = openBytesNative(bytes)
        if (handle == 0L) {
            throw RuntimeException("Failed to open BBF container from bytes")
        }
    }

    override fun close() {
        if (handle != 0L) {
            closeNative(handle)
        }
    }

    val version: Int
        get() = getVersion(handle)

    val headerFlags: Int
        get() = getHeaderFlags(handle)

    val alignment: Int
        get() = getAlignment(handle)

    val reamSize: Int
        get() = getReamSize(handle)

    val assetCount: Long
        get() = getAssetCount(handle)

    val pageCount: Long
        get() = getPageCount(handle)

    val sectionCount: Long
        get() = getSectionCount(handle)

    val metaCount: Long
        get() = getMetaCount(handle)

    fun getMeta(index: Int): BBFMeta {
        val key = getMetaKey(handle, index) ?: ""
        val value = getMetaValue(handle, index) ?: ""
        val parent = getMetaParent(handle, index).takeIf { !it.isNullOrEmpty() }
        return BBFMeta(key, value, parent)
    }

    fun getSection(index: Int): BBFSection {
        val title = getSectionTitle(handle, index) ?: ""
        val startIndex = getSectionStartIndex(handle, index)
        val parent = getSectionParent(handle, index).takeIf { !it.isNullOrEmpty() }
        return BBFSection(title, startIndex, parent)
    }

    fun getPageAssetIndex(index: Int): Long {
        return getPageAssetIndex(handle, index)
    }

    fun getPageFlags(index: Int): Int {
        return getPageFlags(handle, index)
    }

    fun getAsset(index: Int): BBFAsset {
        val offset = getAssetFileOffset(handle, index)
        val size = getAssetFileSize(handle, index)
        val flags = getAssetFlags(handle, index)
        val type = getAssetType(handle, index)
        val hash = getAssetHash(handle, index) ?: ByteArray(16)
        return BBFAsset(offset, hash, size, flags, type)
    }

    fun getAssetData(index: Int): ByteArray? {
        return getAssetData(handle, index)
    }

    fun readAssetDataDirect(index: Int, buffer: ByteBuffer, offset: Int, len: Int): Int {
        if (!buffer.isDirect) {
            throw IllegalArgumentException("ByteBuffer must be a direct buffer")
        }
        return readAssetDataDirect(handle, index, buffer, offset, len)
    }

    fun verifyFooterHash(): Boolean {
        return verifyFooterHash(handle)
    }

    fun verifyAssetHash(index: Int): Boolean {
        return verifyAssetHash(handle, index)
    }

    companion object {
        init {
            System.loadLibrary("bbf")
        }

        @JvmStatic
        private external fun openNative(filePath: String): Long

        @JvmStatic
        private external fun openBytesNative(bytes: ByteArray): Long

        @JvmStatic
        private external fun closeNative(handle: Long)

        @JvmStatic
        private external fun getVersion(handle: Long): Int

        @JvmStatic
        private external fun getHeaderFlags(handle: Long): Int

        @JvmStatic
        private external fun getAlignment(handle: Long): Int

        @JvmStatic
        private external fun getReamSize(handle: Long): Int

        @JvmStatic
        private external fun getAssetCount(handle: Long): Long

        @JvmStatic
        private external fun getPageCount(handle: Long): Long

        @JvmStatic
        private external fun getSectionCount(handle: Long): Long

        @JvmStatic
        private external fun getMetaCount(handle: Long): Long

        @JvmStatic
        private external fun getMetaKey(handle: Long, index: Int): String?

        @JvmStatic
        private external fun getMetaValue(handle: Long, index: Int): String?

        @JvmStatic
        private external fun getMetaParent(handle: Long, index: Int): String?

        @JvmStatic
        private external fun getSectionTitle(handle: Long, index: Int): String?

        @JvmStatic
        private external fun getSectionStartIndex(handle: Long, index: Int): Long

        @JvmStatic
        private external fun getSectionParent(handle: Long, index: Int): String?

        @JvmStatic
        private external fun getPageAssetIndex(handle: Long, index: Int): Long

        @JvmStatic
        private external fun getPageFlags(handle: Long, index: Int): Int

        @JvmStatic
        private external fun getAssetFileOffset(handle: Long, index: Int): Long

        @JvmStatic
        private external fun getAssetFileSize(handle: Long, index: Int): Long

        @JvmStatic
        private external fun getAssetFlags(handle: Long, index: Int): Int

        @JvmStatic
        private external fun getAssetType(handle: Long, index: Int): Int

        @JvmStatic
        private external fun getAssetHash(handle: Long, index: Int): ByteArray?

        @JvmStatic
        private external fun getAssetData(handle: Long, index: Int): ByteArray?

        @JvmStatic
        private external fun readAssetDataDirect(handle: Long, index: Int, buffer: ByteBuffer, offset: Int, len: Int): Int

        @JvmStatic
        private external fun verifyFooterHash(handle: Long): Boolean

        @JvmStatic
        private external fun verifyAssetHash(handle: Long, index: Int): Boolean
    }
}
